### ISSUE TITLE
Q&V: full text search

### ISSUE THREAD

**Issue by urapadmin** [arch-kiosk/arch-kiosk-office#2504]
That's actually the most wanted query option. Ha. No surprises in the age of Google research. 

So I will see if I do this before I move from cool Harris Matrix to cool image annotation.

### a few words on full text search:

It has a linguistic model. So it is searching in a language. That has advantages (it knows that the stem of "digging" is dig and it knows that "and" is a meaningless "stop word" for a search.) and disadvantages (mixing languages is hard, stemming proper nouns, handling of non-words like "FA-12-1"). That is why the results are fuzzy and not algorithmic. 

you can use postfix wildcards but not prefix wildcards. So "FA-012*" is okay, "*-012" isn't. But because of the stemming (reducing words to their linguistic stems) wildcards are a bit odd in the full text search world. 

The hope is, in fact, that it is "good enough" out of the box. That is how we treat google, too. We hardly ever use the options to specify our search and they are limited, anyway.

### syntax of search prompts
"-" results in matches that don't have a certain word. So if you have a bunch of yellow arrows but you want the other colours, you could try _arrow -yellow_ and should get the other ones.
Here is what PostgreSQL tells us about their syntax:

>Quoted word sequences are converted to phrase tests. 
>The word “or” is understood as producing an OR operator, 
>and a dash produces a NOT operator; other punctuation is ignored. This approximates the behavior of some common web search tools.

With "quoted" they mean double quotes. a phrase test means that the words in the quotes must be in exactly that order. That opens up a few questions: 
- words don't have to be written exactly like in the phrase: "cute cats" will find _cute cat_
- stop words are eliminated from phrases. "a very cute cat" finds _cute cat_ because both _very_ and _a_ are stop words. That _very_ has no meaning is debatable in my opinion, but that's what it is.

### the search results
FTS tries to return an excerpt of the matching "document" for every match. A document in our case is a whole record of some type. So e.g. a whole "locus" record is regarded as a single document. That means interpretation, description, type etc. all fields are just part of one document and are being searched. The result is that record. The excerpt is created around the match, or, if there is more than one match in one document, you get a multi-part excerpt connected with "..."). If you think about a long interpretation it will try to give you an excerpt around the match that allows you to identify the match and estimate its relevance for you. So you get a few words before and after the match but rarely the whole document.

**Comment by urapadmin** [comment]
and perhaps I can design this in a way that it can be used in the file repository, too.

**Comment by urapadmin** [comment]
### Kiosk 1.5.27
- very first version of this to evaluate the search results. Full text search is not always the best way to search and often not very intuitive. So it is important to try real data and real queries. 
- The search works like a web search: multiple search words are connected by "and" unless you use "or" explicitly. It supports + and -, as Google does. Currently it fails with single quotation marks which are used to mark complete phrases. 
- full text search is particularly meant for search requests with several words, so testing simply something like "brick" is not all that interesting. 
- the results are ordered by "relevance". I am curious how that goes. Relevance is a very vague term.

**Comment by lbestock** [comment]
Real use is going to need real data, not PVD. I am just very beginning to poke and these are merest initial observations after a long day and some Hamentaschen prep, so not to be taken as real testing. 

- But I am having fun this evening. Very first search was the name of the daughter who was watching me perform the first search, and it was gratifying to be in PVD where something popped up (though the photo is busted, alas). 
- How is it that searching linking+dayplan even gives me link + dayplan? That is unexpected and lovely.
- And disturbed gets me disturbance. This is smarter than I anticipated.
- I expect hitting the return key to perform the search, but it does not.
- The match info highlights the matched words, but when going to the records in which they are found it is not usually easy to find where the search term sits. A highlight there, if possible, would help.
- I did not at first catch that clicking on the context opened that below. When one clicks on the context (might be a bigger Q&V thing) it would be nice if the browser jumped down to that which was just opened. 
- "and" and "or" work as I expect them to, as does +. But I am not sure what - is supposed to do and it doesn't do what the description above led me to expect.

**Comment by urapadmin** [comment]
a few words about full text search:

It has a linguistic model. So it is searching in a language. That has advantages (it knows that the stem of "digging" is dig and it knows that "and" is a meaningless "stop word" for a search.) and disadvantages (mixing languages is hard, stemming proper nouns, handling of non-words like "FA-12-1"). That is why the results are fuzzy and not algorithmic. 

you can use postfix wildcards but not prefix wildcards. So "FA-012*" is okay, "*-012" isn't. But because of the stemming (reducing words to their linguistic stems) wildcards are a bit odd in the full text search world. 

The hope is, in fact, that it is "good enough" out of the box. That is how we treat google, too. We hardly ever use the options to specify our search and they are limited, anyway.

**Comment by urapadmin** [comment]
> * I did not at first catch that clicking on the context opened that below. When one clicks on the context (might be a bigger Q&V thing) it would be nice if the browser jumped down to that which was just opened.

see #2383

**Comment by luizaogs** [comment]
> * The match info highlights the matched words, but when going to the records in which they are found it is not usually easy to find where the search term sits. A highlight there, if possible, would help.

This would indeed be very useful.

**Comment by luizaogs** [comment]
> * "and" and "or" work as I expect them to, as does +. But I am not sure what - is supposed to do and it doesn't do what the description above led me to expect.

Yes, is the - supposed to work like the "or"? It is not.

**Comment by luizaogs** [comment]
Do we need stuff that doesn't match the full text search in the results? Here for instance, do we need to show the type? It seems more intuitive to me to only show the field where the searched for text is found, since one can then click on the identifier and go to the record, but I don't know if there was a specific design choice behind that.

Also some entries include things like type and some don't, and I'm not totally sure why.

<img width="1447" alt="Screenshot 2024-03-24 at 5 33 41 PM" src="https://github.com/arch-kiosk/arch-kiosk-office/assets/67970402/4ab7c4cb-1855-4e61-908a-c5097b2524c9">

**Comment by luizaogs** [comment]
Do you want the match info to be locus_photo, unit_narrative, etc. or locus photo, unit narrative? Or does it not matter? It is currently the former.

**Comment by luizaogs** [comment]
Features, survey units, and excavation units are currently just "unit". Not sure if they should be differentiated.

**Comment by luizaogs** [comment]
Those are minor things, and as far as I can tell the rest is working (though I do think we will be able to see how useful it really is once we have it released to a real project).

**Comment by luizaogs** [comment]
Ah wait. So I searched for "pen" and got these 6 results: 

<img width="1447" alt="Screenshot 2024-03-24 at 5 46 42 PM" src="https://github.com/arch-kiosk/arch-kiosk-office/assets/67970402/a8646c2e-caf9-48ee-a4e9-1384880e9bfe">

<img width="1445" alt="Screenshot 2024-03-24 at 5 46 50 PM" src="https://github.com/arch-kiosk/arch-kiosk-office/assets/67970402/8aa28c39-6830-4118-b7e6-812e9dbeb062">

Then I did LA + pen, expecting to get 5 results (only missing the PBA one), but I get 3: 

<img width="1442" alt="Screenshot 2024-03-24 at 5 47 44 PM" src="https://github.com/arch-kiosk/arch-kiosk-office/assets/67970402/048daa82-926a-48a5-bc72-2f1131e2c217">

But shouldn't the 511-1 records also be found with that search? 511-1 is a collected material in LA: 

<img width="1015" alt="Screenshot 2024-03-24 at 5 48 49 PM" src="https://github.com/arch-kiosk/arch-kiosk-office/assets/67970402/cd879417-158c-4253-b3a5-b9c5ec7a0162">

**Comment by urapadmin** [comment]
> Do you want the match info to be locus_photo, unit_narrative, etc. or locus photo, unit narrative? Or does it not matter? It is currently the former.

This will be covered by #2606

**Comment by urapadmin** [comment]
> > * The match info highlights the matched words, but when going to the records in which they are found it is not usually easy to find where the search term sits. A highlight there, if possible, would help.
> 
> This would indeed be very useful.

- I made a new (not encouraging) ticket for this: #2609

**Comment by urapadmin** [comment]
> > * "and" and "or" work as I expect them to, as does +. But I am not sure what - is supposed to do and it doesn't do what the description above led me to expect.
> 
> Yes, is the - supposed to work like the "or"? It is not.

see my few words on full text search: "-" is supposed to find things WITHOUT the search word. So "+sand -loose" should not come up with loose sand but with - erm, I don't know - other sand.

**Comment by urapadmin** [comment]
> Do we need stuff that doesn't match the full text search in the results? Here for instance, do we need to show the type? It seems more intuitive to me to only show the field where the searched for text is found, since one can then click on the identifier and go to the record, but I don't know if there was a specific design choice behind that.
> 
> Also some entries include things like type and some don't, and I'm not totally sure why.
> <img alt="Screenshot 2024-03-24 at 5 33 41 PM" width="1447" src="https://private-user-images.githubusercontent.com/67970402/316355451-4ab7c4cb-1855-4e61-908a-c5097b2524c9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzNTk3NjAsIm5iZiI6MTcxMTM1OTQ2MCwicGF0aCI6Ii82Nzk3MDQwMi8zMTYzNTU0NTEtNGFiN2M0Y2ItMTg1NS00ZTYxLTkwOGEtYzUwOTdiMjUyNGM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI1VDA5Mzc0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFjNjcyNWI1Y2MzODQyYTAwMjM2ZWFkODU0YjkxNDE1N2JhZjE5N2M0ZmRkMjZhNTI2MmU1NGNlMmJjNDM3ZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ufZq0eNZe8Pa3cR_IfatxLBUpL7lhQuePxiQXdRxKYA">
> 
> Ok so here, for instance. Seems unnecessarily busy when really the match is for a single field:
> <img alt="Screenshot 2024-03-24 at 5 37 47 PM" width="1446" src="https://private-user-images.githubusercontent.com/67970402/316355788-ba40d7c2-65dc-44c1-94c6-7b3e954cfc62.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzNTk3NjAsIm5iZiI6MTcxMTM1OTQ2MCwicGF0aCI6Ii82Nzk3MDQwMi8zMTYzNTU3ODgtYmE0MGQ3YzItNjVkYy00NGMxLTk0YzYtN2IzZTk1NGNmYzYyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI1VDA5Mzc0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlNDk0OGI0ZTA4MTU4YTZmNTBkZTc2ZTY3MjFhODdlYWZmMmQwMjhkY2FmMzEyZWNhZmY5NmZhYzQ0YjVmOGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TB-X816Vh6aT-h_RHpkUcwuWhTPXy332T99_Ds8RuHc">

- see the explanations to search results in the body. 

It really depends if the extra information is useful or not. Google also often gives you stuff that is not necessarily relevant for you in all cases. But in same cases it might help you to decide which of the results is more relevant to you than others. If you search for a single word and get a single search result it appears entirely useless. But not necessarily in other cases. This is a good illustration why I said that results of a full text search are often not so intuitive. 
What I can improve, though, would be adding the labels that give you a hint as to what field originally had the information.
I make a ticket for that, but I think it does not entirely address your concerns: 
-

**Comment by urapadmin** [comment]
we will need a short video tutorial on this that explains a few things about it

**Comment by urapadmin** [comment]
### Kiosk 1.5.29.1
After #2605, this is now available for all projects. So you can test real data with it.

**Comment by luizaogs** [comment]
Can we add this back to testing? Or I will forget about it.

**Comment by urapadmin** [comment]
of course. Back it is.

**Comment by luizaogs** [comment]
I'm trying this in MADP and MKAP and it's working really well. Even relevance -- for instance, in MKAP I just searched for "mixed context" and it gives me the results for that on top, but then later I also get records with "mix" or "mixed", which is exactly what I expected and wanted. 

<img width="1473" alt="Screenshot 2024-04-06 at 9 02 48 AM" src="https://github.com/arch-kiosk/arch-kiosk-office/assets/67970402/c4bbdb5f-8e9a-464d-b298-9c6298d75587">

**Comment by luizaogs** [comment]
Yeah, this looks great as far as I'm concerned. Lutz, let me know if you want me to try anything specific.

**Comment by urapadmin** [comment]
I am glad to hear. Good. And this is what I was waiting for: Using it and testing if it actually serves its purpose in practice. 
Before I close this I would want @lbestock to do the same thing and use it practically on URAP or BUAP.

**Comment by lbestock** [comment]
Whoa. I am playing with this first for real in BUAP and...yeah it's great.

**Comment by lbestock** [comment]
A few times when I have clicked on the identifier in the returned hits list when it is a narrative I have been flown to the top of the unit information, not to the proper narrative. All these narratives I could subsequently get to by scrolling up again and clicking again, but it was a little disconcerting. And now: I think I am finding a pattern, in trying to force it. If the unit has not previously been opened below, then clicking on the identifier just opens the unit, it doesn't fly to the proper narrative. If that unit was open already, then clicking the identifier takes you straight to the right narrative. The unit can be open from a previous search, even, all that seems to matter is that it be previously open.

**Comment by lbestock** [comment]
Not sure if this can be incorporated, but I get different results for First Dynasty and 1st Dynasty, whereas I would want either search to return both.

**Comment by urapadmin** [comment]
> Not sure if this can be incorporated, but I get different results for First Dynasty and 1st Dynasty, whereas I would want either search to return both.

something like that (but that's the problem) can be done with a Thesaurus for the full text search: #2597 But who types all the synonyms in?

**Comment by luizaogs** [comment]
> A few times when I have clicked on the identifier in the returned hits list when it is a narrative I have been flown to the top of the unit information, not to the proper narrative. All these narratives I could subsequently get to by scrolling up again and clicking again, but it was a little disconcerting. And now: I think I am finding a pattern, in trying to force it. If the unit has not previously been opened below, then clicking on the identifier just opens the unit, it doesn't fly to the proper narrative. If that unit was open already, then clicking the identifier takes you straight to the right narrative. The unit can be open from a previous search, even, all that seems to matter is that it be previously open.

Lutz, I think this is probably the issue I had when I couldn’t find a pattern, since as I said above it only happened when I first started testing so by default they were views that weren’t open yet.

**Comment by urapadmin** [comment]
> > A few times when I have clicked on the identifier in the returned hits list when it is a narrative I have been flown to the top of the unit information, not to the proper narrative. All these narratives I could subsequently get to by scrolling up again and clicking again, but it was a little disconcerting. And now: I think I am finding a pattern, in trying to force it. If the unit has not previously been opened below, then clicking on the identifier just opens the unit, it doesn't fly to the proper narrative. If that unit was open already, then clicking the identifier takes you straight to the right narrative. The unit can be open from a previous search, even, all that seems to matter is that it be previously open.
> 
> Lutz, I think this is probably the issue I had when I couldn’t find a pattern, since as I said above it only happened when I first started testing so by default they were views that weren’t open yet.

I have seen it, too. Might be a timing issue. I'll make it its own ticket, though (or is there already one?)

**Comment by luizaogs** [comment]
If those two things are now their own tickets, can we close this?

**Comment by urapadmin** [comment]
ja, it is getting one of those ...
